### ISSUE TITLE
Persistent storage (Feature request and offer to implement)

### ISSUE THREAD

**Issue by derrell** [markabrahams/node-net-snmp#141]
I'm new to node-net-snmp. I'm looking for a way to persist storage of values that are SET so that they can be restored upon system restart, and I don't see any obvious way to do so in the current API. Am I just missing it?

If it does not exist, I'd be happy to implement a generic means of notifying the app of changes such that they can be persisted at the app's option. I'm interested in thoughts on a best implementation.

I'd love to hear ideas! Thanks!

**Comment by markabrahams** [comment]
Hi @derrell - when you register a provider for a scalar or a table, you can optionally specify a "handler" function that does other things you might want to do on each SNMP get or set operation.  See the documentation for the "handler" function in the provider definition of mib.registerProvider().  The handler function is the place under the current design to implement any persistence in response to SNMP protocol actions.  On startup, you could also create a new "mib" instance using snmp.createMib() and then pre-load any data from your persistent storage.  Then pass this "mib" instance into snmp.createAgent() when you create the agent.

So the mechanisms are there for you to roll your own persistent storage solution.

An alternative option is to add support for one or more persistence layers to the library e.g. serialization to file(s), particular database types, etc.  My initial reaction is that this is outside the remit of the library, but I'm happy to hear a persuasive argument for this!  It seems like you're asking for the first solution anyway i.e. generic support for an ability to implement your own persistence, which is already there via the MIB provider handler functions.

**Comment by derrell** [comment]
Hi @markabrahams. Thank you very much for the reply. Yes, I have become intimately familiar with "handler" functions during the past week, as I've implemented each of the dynamic getters, e.g., system uptime. I can certainly see, now, how adding a handler that dealt with persistence. The disadvantage of that approach, as I see it, is that I'm then unable to use the `getProvidersForModule` method and simply register all of them; I'd need to manually create each of the providers. (I suppose I could loop through the `providers` list after calling `getProvidersForModule` and add a handler to each...)

I have come up with a very simple, minimally-invasive change that I believe will work well. I'll simply provide a patch here to show what I'm thinking. I'll be testing it this week, and if all goes as expected, I'll provide a PR for your consideration. Here's what I'm thinking:

```
diff --git a/index.js b/index.js
index c9be987..0cab629 100644
--- a/index.js
+++ b/index.js
@@ -3518,6 +3518,8 @@ var Mib = function () {
 	this.providerNodes = {};
 };
 
+util.inherits (Mib, events.EventEmitter);
+
 Mib.prototype.addNodesForOid = function (oidString) {
 	var address = Mib.convertOidToAddress (oidString);
 	return this.addNodesForAddress (address);
@@ -3765,6 +3767,7 @@ Mib.prototype.setScalarValue = function (scalarName, newValue) {
 		instanceNode.valueType = provider.scalarType;
 	}
 	instanceNode.value = newValue;
+	this.emit("setScalarValue", instanceNode, provider);
 	// return instanceNode.setValue (newValue);
 };
 
@@ -3907,6 +3910,7 @@ Mib.prototype.addTableRow = function (table, row) {
 		instanceNode = this.lookup (instanceAddress);
 		instanceNode.valueType = column.type;
 		instanceNode.value = row[rowValueOffset + i];
+		this.emit("addTableRow", instanceNode, provider);
 	}
 };
 
@@ -4014,6 +4018,7 @@ Mib.prototype.setTableSingleCell = function (table, columnNumber, rowIndex, valu
 	columnNode = providerNode.children[columnNumber];
 	instanceNode = columnNode.getInstanceNodeForTableRowIndex (instanceAddress);
 	instanceNode.value = value;
+	this.emit("setTableSingleCell", instanceNode, providerNode.provider);
 	// return instanceNode.setValue (value);
 };
 
@@ -4226,11 +4231,11 @@ Agent.prototype.onMsg = function (buffer, rinfo) {
 		this.callback (new RequestInvalidError ("Unexpected PDU type " +
 			message.pdu.type + " (" + PduType[message.pdu.type] + ")"));
 	}
-
 };
 
 Agent.prototype.request = function (requestMessage, rinfo) {
 	var me = this;
+	var mib = this.mib;
 	var varbindsCompleted = 0;
 	var requestPdu = requestMessage.pdu;
 	var varbindsLength = requestPdu.varbinds.length;
@@ -4307,6 +4312,7 @@ Agent.prototype.request = function (requestMessage, rinfo) {
 				} else {
 					if ( requestPdu.type == PduType.SetRequest ) {
 						mibRequests[savedIndex].instanceNode.setValue (requestPdu.varbinds[savedIndex].value);
+						mib.emit("agentSetValue", instanceNode, providerNode.provider);
 					}
 					if ( ( requestPdu.type == PduType.GetNextRequest || requestPdu.type == PduType.GetBulkRequest ) &&
 							requestPdu.varbinds[savedIndex].type == ObjectType.EndOfMibView ) {
```

Thanks again, and your thoughts on this approach are appreciated!

Derrell

**Comment by markabrahams** [comment]
Hi @derrell - that's a good idea, and a much more "modern" approach to the problem.

A few thoughts on what events to include in an EventEmitter implementation:
(1) I wonder if you should emit on both SNMP protocol operations and library events (as you've proposed), or on protocol operations only?
(2) If you're emitting on library events, then you might consider emitting on other "change" operations e.g. row deletion?
(3) Although your use case (persistence) requires emitting on `set` operations only, you might consider emitting events for `get` protocol operations as well, as others cases may want to receive events on these?  If you're emitting on library events, then for consistency - if you went this way - you'd be emitting on the library "get" calls (e.g. `getTableSingleCell`, `getTableRowCells`).

Having said that, the reservation I have with turning the Mib into an EventEmitter is that you'll notice that the entire library is based on the now somewhat dated callback mechanism for asynchronous operations (the library is coming on 8 years old!).  I did seriously contemplate converting the entire library from callbacks to promises earlier this year, but through a series of circumstances, this didn't happen.  An EventEmitter would stick out like a sore thumb a bit if used in just this isolated instance, similar to converting a single library call to returning a promise, but leaving the rest of the library to use callbacks.  So there's a internal library consistency argument for the status quo of callbacks, unless you're making a more wholesale change across the whole library.

Also, I have a slight reservation of having two asynchronous mechanisms in play for some of the same thing (SNMP protocol operations will have two async handling mechanisms).  There's a good aspect of providing such a choice, but also a downside of some possible extra confusion for the library user, and the burden to discover how both mechanisms work, determine what specific operations invoke each one, and then decide which is the best for them.

So although I'm not totally convinced at this point, I'm still open to being convinced!

**Comment by derrell** [comment]
Thinking about your comments...

I think that maybe there's no reason to notify the app of library events, as the app is already issuing those. Only protocol events should be necessary. And sure, although it's likely that the app will be most interested in SET operations, it may want to monitor other operations as well, so I can alert the app to all protocol operations.

This was not the first use of an event emitter in the library -- Session is also an event emitter -- but I can as easily have the app register a callback function, as have the app await events. What I may do is emit an event, internally (which can later be documented as part of the public interface) but listen to that event and call the registered callback function, as the documented public interface, for now.

**Comment by markabrahams** [comment]
Yes - you're right - Session is an EventEmitter!  (I didn't write that code, so I'd forgotten about it!)  As such, that's removed a barrier to making the Agent or Mib an EventEmitter.

And also agree on protocol operations only, not library functions.  You're calling the library functions already, so you already have all the context you need for these.

So you've put all but one of my reservations to bed!  The only outstanding one is my musing above about whether to have the two mechanisms for largely the same thing.  Any convincing argument there?!? :-)

I think looping through providers and adding handlers isn't a heinous burden for the user.  Further, to allay this, we could add a handler function as an optional second argument to `getProvidersForModule()` and have the handler added to every provider created as a convenience gesture?  Thoughts?

**Comment by derrell** [comment]
I'm open to this. And for my persistence purpose, it would work fine. However, after adding all of the events for other agent operations, including error events, there will be some cases where we can't notify the app if it's all handled in the provider. Some of the errors occur when no provider is known, so those could not be reported tot he app. My feeling is that we should either provide all relevant information, as I've tried to do, or the feature is far less useful. An app that wanted to monitor requests to its agent would not be able to do so effectively if requests that generated certain types of errors couldn't be reported to the app. If we went this direction, I'd be inclined to handle _only_ the events required for persistence, which all occur after the provider is available...

**Comment by markabrahams** [comment]
Error handling is already done via the Agent's callback - a mandatory parameter in the snmp.createAgent() call.  This is the current signalling mechanism for agent errors outside of providers to the app.

On further inspection, it was I who introduced the inconsistency, which I generally try quite hard to avoid! :-(  `Session` doesn't register a general callback, but emits events for general errors outside of errors on specific API calls (e.g. session.getNext), which do use a callback supplied to the specific `Session` API call for errors pertaining to that call.  I implemented the `Agent`, and did the opposite - there's a general callback used to signal a whole bunch of error conditions, and I didn't use an EventEmitter for this.

Even though an EventEmitter is a superior solution, switching over to this wholesale (i.e. removing the global `Agent` callback) would be quite some breaking change, which is always possible, but something this big I have less of a stomach for the havoc it might cause!  And introducing the EventEmitter without removing the global `Agent` callback invites confusion for library implementers and library users alike on what mechanism to use when, with the two doing largely the same thing.

**Comment by derrell** [comment]
On further investigation, it appears that the standard provider handler can't be used for persistence. The handler would need intimate knowledge of the internal workings of the library, regarding any automatic actions that will be or were accomplished as a result of "read-create" or RowStatus functionality. The handler sees, for example, a MibRequest for a RowStatus column update for a (formerly) non-existent row, doesn't know that the row was just created (or what the default values are), or even that the value that was specified for that RowStatus column, e.g., "createAndGo", has been changed to "active". I think that's way too much required knowledge and implementation to push to the user.

I see two viable choices (not considering event emitting for the moment): a global agent event callback function, as is currently pushed to the `agent-events` branch; or the addition of an `agentEventHandler` to the provider along with an added parameter to `getProvidersForModule` that passes a common handler for that purpose. My preference is the former, since I see the latter as most likely simply consuming memory in every provider object without any real benefit. I believe it is unlikely that different `agentEventHandler` functions will be provided to different providers, although I suppose it is possible.

There could, and likely will be, multiple calls to the agent event handler, for a single agent request. For example, for the request to set a RowStatus column to `createAndGo` (4) using this command:

```
snmpset -v2c -c private localhost:1611 ifEntry.99.7 i 4
```

this output is generated with console.log statements in the agent event handler and the callback functions:

```
Agent event:  {
  eventType: 'autoCreateTableRow',
  oid: '1.3.6.1.2.1.2.2.1.99.7',
  providerName: 'ifTable',
  values: [ 7, 'Hello world, amended!', 2, 1 ]
}
Agent event:  {
  eventType: 'set',
  requestType: 'SetRequest',
  oid: '1.3.6.1.2.1.2.2.1.99.7',
  providerName: 'ifTable',
  value: 1
}
Agent event:  {
  eventType: 'get',
  requestType: 'SetRequest',
  oid: '1.3.6.1.2.1.2.2.1.99.7',
  providerName: 'ifTable',
  value: 1
}
callback: [
  {
    "oid": "1.3.6.1.2.1.2.2.1.99.7",
    "type": 2,
    "value": 1
  }
]
```

Am I off base here?

**Comment by derrell** [comment]
Documentation has been added, for the current implementation

**Comment by markabrahams** [comment]
Hi @derrell - now that we have a `createHandler` on a provider as part of the row-status PR, that *could* be used to address your concerns with doing some persistence by just the provider `handler` alone.

Having said that, I'm warming to the idea of an agent EventEmitter a bit more.  My main concern now - as mentioned above - is providing two mechanism to support similar things.  The reason I'm warming to it is that I can see that the `createHandler` / `handler` callbacks are well-positioned to intercept protocol actions and dynamically update the MIB where necessary.  The emits are positioned more for reporting the results of the MIB/protocol events, so in my mind there is some distinction for these being positioned for separate duties.  Of course they're both quite general and could be used in any way, but that helps somewhat justify their coexistence.

For the above example, I wonder if a SetRequest should only fire a single event, with all the relevant details on it?  If it does the "set" and "get" separate events, wouldn't you have the "createAndGo" (4) value on the "set" event?

Anyway, I'll get into a review of the code for this hopefully tomorrow to try to keep up with you!  (lots going on at the moment and it's a bit late!)

**Comment by derrell** [comment]
Although my current implementation uses an event emitter internally, it listens on the events internally and calls a registered callback function, to keep the callback style in place. I did it that way so that in the future, when event emitters are more appropriate, that change is trivial.

Allowing for event reporting in the existing handler would likely require some significant changes to MibRequest. The biggest problem with that is that unchanged code would expect to be able to process any call to its handler, yet there would now be new calls to the handler that it needn't care about. I don't see the justification for trying to push this into the existing handler, at the expense of a major version bump and necessary code changes for users.

Initially, I considered a single event (regardless of to where it is sent), but it would require passing a queue of changes around to various functions, and didn't work well. Also, each of the current events that is sent can be processed independently, so I've found, in actually using this, that there would be no great benefit in coalescing events. I've implemented persistent storage in my app, using the current implementation, and it's working well and was easy to implement.

Anyway, I'll await your review and comments. Thanks again.

**Comment by markabrahams** [comment]
Hi @derrell - I've reviewed at a high-level, and am currently contemplating the event quantum and format.  The options I'm thinking about are:
(a) each create, request and response action generates a separate event (your current implementation)
(b) a single event is generated for the the create, request and response actions of a single varbind
(c) a single event is generated for all varbinds in a single PDU i.e. an array of items in (b).

I imagine there will be cases where the library user would want to have all of the create/request/response information for a single varbind set/get together when using the event data.  In this case, it would be easier for the user to get a single event for this ala (b) above, otherwise they'd need some sort of matching / caching to correlate create, request and response events for use together.  Obviously for your case you don't need this, but some might, so something to consider.

There's even a case for having a single event representing the outcome of an entire PDU - option (c) above.  This is attractive to give accurate information if the library ever implements the two-phase SetRequest validation/commit ala RFC 3416 Section 4.2.5 (it doesn't at the moment, but I'm thinking I will probably do this at some point), as if there's a 2nd-phase failure it's a nicer communication for the library to undo the previous sets silently and then just send an event representing the resulting error, rather than incrementally sending what ultimately will be incorrect (rolled-back) results.

My thinking here is slanted towards the model that caters for ease of library use and most comprehensive use-case coverage, which I think is (b) or (c), even at the expense of a more complex implementation, although I *think* I can envision an implementation of (b) or (c) that isn't *too* nasty.

I know it would be a reasonable divergence from your current design, but wondering if there's any merit in it?

The other thing at a high-level that could help an implementation of (b) or (c) is to store any information for the emitted event on the mibRequest along the way (we can stack up more info on the mibRequest no problems, as any additional info could be useful in the `handler` anyway), and then just have one piece of code that produces a varbind's event from the mibRequest in its final state, and emits that event, rather than having the events emitted all the way through processing the request.

I might do an implementation of (b) or (c) to get a feel for that approach.  Sometimes it's not until implementation that I realise some of the pros and cons (more likely the cons!) of what might seem like a good idea to me on first consideration!

Final thought is that there's already an agent-global `callback` that I'd probably use for this in the typical `function (error, data)` callback format (currently it only gets errors, but since that's the first argument, no harm in extending to data as well).  Just to save a fourth user-feedback / callback mechanism (`createHandler`, `handler`, `callback` and `agentEventHandler`).  Having said that - I'm really swaying towards the agent as an EventEmitter, and replacing the agent's `callback` function altogether with this.

**Comment by derrell** [comment]
@markabrahams Ask and ye shall receive. :-)

I've updated it so the developer can now select between three style(s) of passing the events to the user:
- event (EventEmitter)
- mainCallback (the callback registered at agent instantiation time)
- separateCallback (the way I had implemented it previously, with a registered callback for events

The developer may also select when events are generated to the user via the above-selected style(s):
- individual (each event is sent separately)
- coalesced (events are collected and sent at the end of processing a pdu)

If events are desired, the developer would select one or more styles, and one or more _whens_. In the example code, I demonstrate selecting all of them. A real application would likely select one from the styles group, and one from the when group.

**Comment by markabrahams** [comment]
Wow @derrell - it's a smorgasbord of options!  Thanks for doing that, as my schedule has been a bit full over the last few days so I didn't get to an implementation like I intended.  It's getting a bit late (early!) here now, but I'll review this tomorrow.

**Comment by derrell** [comment]
Thanks @markabrahams. This looks like it will almost do everything necessary. The only thing I've found so far that's missing is that when setting a single column of a table row, the row index and column are missing. Here's an example. After having created a new row, using
```
snmpset -v2c -c private localhost:1611 ifTable.1.99.3 i 4
```
I then set a column value with
```
snmpset -v2c -c private localhost:1611 ifTable.1.2.3 s 'hi there'
```
The resulting varbind at the callback yielded:
```
[
  {
    "oid": "1.3.6.1.2.1.2.2.1.2.3",
    "type": 4,
    "value": {
      "type": "Buffer",
      "data": [
        104,
        105,
        32,
        116,
        104,
        101,
        114,
        101
      ]
    },
    "requestType": 4,
    "requestValue": {
      "type": "Buffer",
      "data": [
        104,
        105,
        32,
        116,
        104,
        101,
        114,
        101
      ]
    },
    "providerName": "ifTable"
  }
]
```
I suggest that there be `rowIndex` and `column` fields. For my purposes, the latter would be the index into tableColumns, not the "number", although I suppose some users would be interested in the number, so maybe add both `columnIndex` and `columnNumber` fields?

Here are my yet-untested changes to snmp-agent.js:

```
diff --git a/example/snmp-agent.js b/example/snmp-agent.js
index 2c1c23e..a6bc9c7 100644
--- a/example/snmp-agent.js
+++ b/example/snmp-agent.js
@@ -12,12 +12,80 @@ var snmpOptions = {
     accessControlModelType: snmp.AccessControlModelType.Simple
 };
 
+var changes;
+
+// If there's a persistent store, make its specified changes
+try {
+	changes = JSON.parse(fs.readFileSync("persistent.json"));
+
+	for (var providerName in changes) {
+		var change = changes[providerName];
+		if (typeof change == "object") {
+			// table row
+			for (var rowIndex in change) {
+				mib.addTableRow(providerName, change[rowIndex]);
+			}
+		} else {
+			mib.setScalarValue(providerName, change);
+		}
+	}
+} catch (e) {
+	console.log("Could not parse persistent storage");
+	changes = {};
+}
+
+
 var callback = function (error, data) {
-    if ( error ) {
-        console.error (error);
-    } else {
-        console.log (JSON.stringify(data.pdu.varbinds, null, 2));
-    }
+	let index;
+
+	if ( error ) {
+		console.error (error);
+		return;
+	}
+
+	console.log (JSON.stringify(data.pdu.varbinds, null, 2));
+
+	data.pdu.varbinds.forEach(
+		(varbind) =>
+		{
+			let value;
+
+			if (varbind.autoCreated && varbind.autoCreatedRowIndex)
+			{
+				// Auto-create table row
+				index = JSON.stringify(varbind.autoCreatedRowIndex);
+				if ( ! changes[varbind.providerName] ) {
+					changes[varbind.providerName] = {};
+				}
+				value = varbind.value;
+				value = value.map((v) => v instanceof Buffer ? v.toString() : v);
+				changes[varbind.providerName][index] = value;
+			} else if (varbind.autoCreated) {
+				// Auto-created scalar
+				value = varbind.value;
+				value = value instanceof Buffer ? value.toString() : value;
+				changes[varbind.providerName] = value;
+			} else if (varbind.deleted && varbind.deletedRowIndex) {
+				// Delete table row
+				index = JSON.stringify(varbind.deletedRowIndex);
+				if (changes && changes[varbind.providerName] && changes[varbind.providerName][index]) {
+					delete changes[varbind.providerName][index];
+					if (Object.keys(changes[varbind.providerName]).length === 0) {
+						delete changes[varbind.providerName];
+					}
+				}
+			} else if (varbind.rowIndex && "column" in varbind) {
+				// Set a column value
+				value = varbind.value;
+				changes[varbind.providerName][varbind.rowIndex].value[varbind.column] =
+					value instanceof Buffer ? value.toString() : value;
+			} else {
+				// Set a scalar
+				value = varbind.value;
+				changes[varbind.providerName] =
+					value instanceof Buffer ? value.toString() : value;
+			}
+		});
 };
 
 var agent = snmp.createAgent(snmpOptions, callback);
```
Note that in the `// Set a column value` section, I expect the currently non-existent `varbind.rowIndex` and `varbind.column` members to be present.

**Comment by derrell** [comment]
Actually, much better than just `rowIndex` and `column` when a column's value is changed, also include the whole row data array in the event, as is done with `autoCreated: true` and `deleted: true`.

**Comment by markabrahams** [comment]
Thanks @derrell for accommodating this!  Yep - great idea - I'll add `rowIndex` and `column` for gets and sets on a table cell, and a `changedRow` for sets on a cell.

**Comment by markabrahams** [comment]
OK I've merged the `callback-status` branch into master now and published as verison 3.2.0 of the npm.

A minor change to the above discussion: Row creation, deletion, normal cell set and get all get standard fields `rowIndex`, `row`, and `column` (instead of `deletedRow`, `createdRow` etc.).  So use the `autocreated: true` and `deleted: true` to detection creations and deletions.  Was just tidier for development and hopefully users to have these fields as standardised names.

I haven't had a chance to update the README with usage info, so at the moment there are a lot of undocumented fields in the callback varbind info - will hopefully get to that shortly!

Thanks again @derrell for your patience with this one!  My testing has checked out, but let me know if there are any issues.

**Comment by derrell** [comment]
Very cool... but not quite right. My very first test crashed it. Try running `example/snmp-agent.js` and then:
```
snmpget -v2c -c private localhost:1611 snmpEnableAuthenTraps.0
```

The problem is that this is a `read-create` object, so it gets created ty `tryCreateInstance`, but then at index.js:4886, it assumes that if `tryCreateInstance` returned something, it must be a table row creation, but that's not a correct assumption.

**Comment by derrell** [comment]
On this request:
```
snmpset -v2c -c private localhost:1611 sysDescr.0 s 'hi there'
```
it crashes on index.js:4839 because `provider` is dereferenced, but `provider` has only been set if it's a Table.

**Comment by markabrahams** [comment]
Oops - that's not good!  I've pushed a fix to master for those two bugs and released the fix in version 3.2.1 of the npm.

**Comment by derrell** [comment]
Great. I'll do some more testing. I did notice one other thing: The callback gets called on GETs as well as SETs. I thought I could distinguish between them using `data.pdu.type`, but it is GetResponse regardless of whether the initial operation was a GET or a SET. For persistence, I'd want to ignore GETs. Can you add some indication in the callback `data` parameter that indicates the initial request?

**Comment by derrell** [comment]
> I did notice one other thing: The callback gets called on GETs as well as SETs. I thought I could distinguish between them using `data.pdu.type`, but it is GetResponse regardless of whether the initial operation was a GET or a SET. For persistence, I'd want to ignore GETs. Can you add some indication in the callback `data` parameter that indicates the initial request?

Disregard. I think all of the necessary information is available. I'll repost if I find otherwise.

**Comment by derrell** [comment]
There is an inconsistency that should possibly be corrected. On an auto-created row (RowStatus column set to 4), the callback varbind data has the `row` member with the post-operation data, i.e., it contains the row with all of the default values filled in. However, on setting the value of a specific column of an existing row, the callback varbind data has the `row` member with the pre-operation data, i.e., without that specific column's newly-set value. It is possible to generate the post-operation data with the information provided, with `varbind.row[varbind.columnPosition] = varbind.value;` and it may be that there's a time that having the old data in the event is useful, although I'm not convinced there is. The inconsistency between the two operations is what bothers me.

Thoughts?

**Comment by derrell** [comment]
I would probably prefer to see a new member added to each of these set request events, `oldValue`, and then have `row` always contain the complete new row, including the changed `value`.

**Comment by markabrahams** [comment]
Yes - I agree on your take on the inconsistency and I like your solution of including an `oldValue` and the `row` always being the final state of the row.  Feel free to create a PR for this, or I can do this in the next day or so.

**Comment by markabrahams** [comment]
I've made the above `oldValue` and `row` changes, and also converted buffer values to string values, which will take away the need for the library user to be casting buffers to strings all the time.

Changes on master now and published in version 3.4.1 of the npm.

**Comment by derrell** [comment]
Cool! Thanks!
